### PR TITLE
`General`: Add SAML2 audit features

### PR DESCRIPTION
**Pull request by dfuchss**
⚠️ This is only testable with SAML2 Configuration; It will not influence other deployments by design ⚠️ 

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

The audit log does not contain the human-readable login information if you log in via SAML2. See #9228 

### Description

Adds SAML2 Events to Audit Log
image

### Steps for Testing

Prerequisites:
- SAML2 Instance
- Test Login with Password
- Test Login with SAML2
- Test Account creation

You should find all related information now readable without duplicates in the AuditLog.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

## Summary by CodeRabbit

- **New Features**
	- Introduced support for SAML2 authentication, enhancing the security and service integration capabilities of the application.
	- Added detailed audit logging for user authentication events to improve tracking and accountability.

- **Bug Fixes**
	- Updated configuration properties and annotations for SAML2 management, ensuring consistency and reducing potential errors in configuration.
	
- **Documentation**
	- Enhanced code documentation to clarify the purpose and usage of new constants and configurations related to SAML2.

### CODE REVIEW

**Review by b-fein — DISMISSED** [review]
Code changes make sense. (Testing later when I find time for it.)

Just to be sure: The screenshot in the description still shows a base64-encoded+encrypted username. Was this just an intermediate version during testing or are there duplicate events now, once with encoded username and once without?

**Comment by dfuchss** [comment]
> Code changes make sense. (Testing later when I find time for it.)
> 
> Just to be sure: The screenshot in the description still shows a base64-encoded+encrypted username. Was this just an intermediate version during testing or are there duplicate events now, once with encoded username and once without?

The problem is that I could try to filter them similar to [`1654b27` (#9233)](https://github.com/ls1intum/Artemis/pull/9233/commits/1654b273dbe257fce0b9688db212275ba56a6bed) (in the mentioned commit, the type is wrong, but the idea is the same). The problem is that this would mean that I need the information whether SAML2 is active, I.e., I'd need to use a service in a repository. That leads to an architecture violation. Therefore, I reverted that implementation. Currently, the encrypted is shown as well. If you have an idea how to access the active profile without the use of this service, just tell it to me. Then I can add the filtering as well :)

**Comment by dfuchss** [comment]
@b-fein now it should work :D -> https://github.com/ls1intum/Artemis/pull/9233/commits/b77f1ca371d3fdc20ad74d8a2bd86191e6a1f47e

File: src/main/java/de/tum/in/www1/artemis/repository/CustomAuditEventRepository.java [hunk 1]
```diff
@@ -62,6 +78,10 @@ public void add(AuditEvent event) {
         }
     }
 
+    private boolean isSaml2Active() {
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/repository/CustomAuditEventRepository.java:81]: See commit-comment: https://github.com/ls1intum/Artemis/commit/b77f1ca371d3fdc20ad74d8a2bd86191e6a1f47e#r145547501

**Review by b-fein — APPROVED** [review]

**Review by b-fein — DISMISSED** [review]
Also works when testing on a local test system.

**Review by SimonEntholzer — APPROVED** [review]
Code changes make sense :+1:

**Review by Strohgelaender — COMMENTED** [review]
looks good, just one small comment:

File: src/main/java/de/tum/in/www1/artemis/repository/CustomAuditEventRepository.java [hunk 2]
```diff
@@ -37,9 +45,10 @@ public class CustomAuditEventRepository implements AuditEventRepository {
 
     private static final Logger log = LoggerFactory.getLogger(CustomAuditEventRepository.class);
 
-    public CustomAuditEventRepository(PersistenceAuditEventRepository persistenceAuditEventRepository, AuditEventConverter auditEventConverter) {
+    public CustomAuditEventRepository(Environment environment, PersistenceAuditEventRepository persistenceAuditEventRepository, AuditEventConverter auditEventConverter) {
         this.persistenceAuditEventRepository = persistenceAuditEventRepository;
         this.auditEventConverter = auditEventConverter;
+        this.isSaml2Active = Set.of(environment.getActiveProfiles()).contains(Constants.PROFILE_SAML2);
```
[the remark this answers is not in the document]
Comment (dfuchss) [reply src/main/java/de/tum/in/www1/artemis/repository/CustomAuditEventRepository.java:51]: See https://github.com/ls1intum/Artemis/actions/runs/10460146560/job/28965706676?pr=9233

[the remark this answers is not in the document]
Comment (dfuchss) [reply src/main/java/de/tum/in/www1/artemis/repository/CustomAuditEventRepository.java:51]: It's not allowed to use a Service in a Repository .. thus, this was the only option without modifying the tests directly.

**Review by Strohgelaender — DISMISSED** [review]
code lgtm

**Review by BaumiCoder — APPROVED** [review]
Approve Code :+1:

**Review by b-fein — APPROVED** [review]

**Comment by dfuchss** [comment]
@BaumiCoder @Strohgelaender @SimonEntholzer could you also re-approve (if I didn't make an error during merge :D )

**Review by BaumiCoder — APPROVED** [review]

**Review by b-fein — APPROVED** [review]

**Review by SimonEntholzer — APPROVED** [review]
Re-approve code

**Review by BaumiCoder — APPROVED** [review]

**Review by Strohgelaender — APPROVED** [review]

**Review by krusche — APPROVED** [review]

**Pull request merged**